### PR TITLE
integrate with crypto module to decrypt password

### LINKED ISSUE #1983: Integrate Datasource client with credential manager to decrypt

### PR DESCRIPTION
**Pull request by zhongnansu**
Signed-off-by: Zhongnan Su <szhongna@amazon.com>

### Description
integrate with crypto module to decrypt password
 
### Issues Resolved
#1983
 
### Check List
- [ ] New functionality includes testing.
  - [ ] All tests pass
    - [ ] `yarn test:jest`
    - [ ] `yarn test:jest_integration`
    - [ ] `yarn test:ftr`
- [ ] New functionality has been documented.
- [ ] Commits are signed per the DCO using --signoff

### CODE REVIEW

**Review by noCharger — APPROVED** [review]
LGTM. I will probabaly refactor CryptographyClient  to be a service

File: src/plugins/data_source/server/client/configure_client.ts [hunk 1]
```diff
@@ -11,23 +11,28 @@ import {
   SavedObjectsErrorHelpers,
 } from '../../../../../src/core/server';
 import { DATA_SOURCE_SAVED_OBJECT_TYPE, CREDENTIAL_SAVED_OBJECT_TYPE } from '../../common';
-import { CredentialSavedObjectAttributes } from '../../common/credentials/types';
+import {
+  CredentialSavedObjectAttributes,
+  UsernamePasswordTypedContent,
+} from '../../common/credentials/types';
```
Comment (noCharger) [inline src/plugins/data_source/server/client/configure_client.ts:17]: `../../common` works as well

File: src/plugins/data_source/server/client/configure_client.ts [hunk 2]
```diff
@@ -48,13 +53,24 @@ export const getDataSource = async (
...
     );
+    const {
+      username,
+      password,
+    } = credentialSavedObject.attributes.credentialMaterials.credentialMaterialsContent;
+    const decodedPassword = await cryptographyClient.decodeAndDecrypt(password);
+    const credential = {
+      username,
+      password: decodedPassword,
+    };
```
Comment (noCharger) [inline src/plugins/data_source/server/client/configure_client.ts:72]: How about
```
const credential = {
      username,
      password: await cryptographyClient.decodeAndDecrypt(password),
 };
```

File: src/plugins/data_source/server/data_source_service.ts [hunk 3]
```diff
@@ -6,11 +6,13 @@
 import { Logger, OpenSearchClient, SavedObjectsClientContract } from '../../../../src/core/server';
 import { DataSourcePluginConfigType } from '../config';
 import { OpenSearchClientPool, configureClient } from './client';
+import { CryptographyClient } from './cryptography';
 export interface DataSourceServiceSetup {
   getDataSourceClient: (
     dataSourceId: string,
     // this saved objects client is used to fetch data source on behalf of users, caller should pass scoped saved objects client
-    savedObjects: SavedObjectsClientContract
+    savedObjects: SavedObjectsClientContract,
```
Comment (noCharger) [inline src/plugins/data_source/server/data_source_service.ts:14]:
```
    // this saved objects client is used to fetch data source on behalf of users, caller should pass scoped saved objects client
```
Can I ask more context on this?

**Review by noCharger — APPROVED** [review]

File: src/plugins/data_source/server/client/configure_client.ts (hunk 1 above)
Comment (zhongnansu) [reply src/plugins/data_source/server/client/configure_client.ts:17]: image

**Comment by zhongnansu** [comment]
CI failed due to #2169 
Merging in to unblock future PR

File: src/plugins/data_source/server/data_source_service.ts (hunk 3 above)
Comment (zhongnansu) [reply src/plugins/data_source/server/data_source_service.ts:14]: it should be scoped client, which means a client tight to a user in the request. Different user may have access to different saved object, including datasource or credentials.

Comment (noCharger) [reply src/plugins/data_source/server/data_source_service.ts:14]: > it should be scoped client, which means a client tight to a user in the request. Different user may have access to different saved object, including datasource or credentials.

Follow up question - Is this already possible with the current scoped clienet? What is the isolation based on?

File: src/plugins/data_source/server/client/configure_client.ts (hunk 1 above)
Comment (noCharger) [reply src/plugins/data_source/server/client/configure_client.ts:17]: i will refactor on this to expose directly

Comment (zengyan-amazon) [reply src/plugins/data_source/server/client/configure_client.ts:17]: let's create an issue to track this

**Pull request merged**